_posts/2020-10-19-torch-image-classification/torch-image-classification.Rmd

```diff
@@ -0,0 +1,543 @@
+
+```{r}
+train_dl$.length() 
+valid_dl$.length() 
+test_dl$.length()  
+```
+
+    490
+    18
+    18
+
+#### Some birds
+
+Next, let's view a few images from the test set. We can retrieve the first batch -- images and corresponding classes -- by creating an iterator from the `dataloader` and calling `next()` on it:
+
+```{r}
+# for display purposes, here we are actually using a batch_size of 24
+batch <- train_dl$.iter()$.next()
+```
+
+`batch` is a list, the first item being the image tensors:
+
+```{r}
+batch[[1]]$size()
+```
+
+    [1]  24   3 224 224
+
+And the second, the classes:
+
+```{r}
+batch[[2]]$size()
+```
+
+    [1] 24
+
+Classes are coded as integers, to be used as indexes into a vector of class names. We'll use those for labeling the images.
+
+```{r}
+classes <- batch[[2]]
+classes
+```
+
+    torch_tensor 
+     1
+     1
+     1
+     1
+     1
+     2
+     2
+     2
+     2
+     2
+     3
+     3
+     3
+     3
+     3
+     4
+     4
+     4
+     4
+     4
+     5
+     5
+     5
+     5
+    [ GPULongType{24} ]
+
+The image tensors have shape `batch_size x num_channels x height x width`. For plotting using `as.raster()`, we need to reshape the images such that channels come last. We also undo the normalization applied by the `dataloader`.
+
+Here are the first twenty-four images:
+
+```{r, fig.width = 12, fig.height = 8}
+library(dplyr)
+
+images <- as_array(batch[[1]]) %>% aperm(perm = c(1, 3, 4, 2))
+mean <- c(0.485, 0.456, 0.406)
+std <- c(0.229, 0.224, 0.225)
+images <- std * images + mean
+images <- images * 255
+images[images > 255] <- 255
+images[images < 0] <- 0
+
+par(mfcol = c(4,6), mar = rep(1, 4))
+
+images %>%
+  purrr::array_tree(1) %>%
+  purrr::set_names(class_names[as_array(classes)]) %>%
+  purrr::map(as.raster, max = 255) %>%
+  purrr::iwalk(~{plot(.x); title(.y)})
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-page"}
+knitr::include_graphics("images/image_classif_birds.png")
+```
+
+## Model
+
+The backbone of our model is a pre-trained instance of Resnet.
+
+```{r}
+model <- model_resnet18(pretrained = TRUE)
+```
+
+But we want to distinguish between our 225 bird species, while ResNet was trained on 1000 different classes. What can we do? We simply replace the output layer.
+
+The new output layer is also the only one whose weights we are going to train -- leaving all other ResNet parameters the way they are. Technically, we *could* perform backpropagation through the complete model, striving to fine-tune Resnet's weights as well. This, however, would slow down training significantly. In fact, the choice is not all-or-none: It is up to us how many of the original parameters to keep fixed, and how many to "set free" for fine tuning. For the task at hand, we'll be content to just train the newly added output layer: With the abundance of animals, including birds, in ImageNet, we expect the trained ResNet to know a lot about them!
```

Comment:
```suggestion
The new output layer is also the only one whose weights we are going to train -- leaving all other ResNet parameters the way they are. Technically, we *could* perform backpropagation through the complete model, striving to fine-tune ResNet's weights as well. However, this would slow down training significantly. In fact, the choice is not all-or-none: It is up to us how many of the original parameters to keep fixed, and how many to "set free" for fine tuning. For the task at hand, we'll be content to just train the newly added output layer: With the abundance of animals, including birds, in ImageNet, we expect the trained ResNet to know a lot about them!
```

Reply:
If there's a pattern to the ResNet vs Resnet I'm not seeing it… 😬